### PR TITLE
test(imports): add concurrency and multi-file cycle tests for ImportsValidator (#456)

### LINKED ISSUE #456: Add concurrency tests for ImportsValidator circular dependency detection
## Category
Test Quality

## Severity
Critical

## Location
`crates/agnix-core/src/rules/imports.rs`

## Description
The ImportsValidator handles complex graph traversal with circular dependency detection and depth limits, but existing tests only cover basic happy-path scenarios. Critical error paths are untested:

1. No tests for circular chains involving 3+ files (A->B->C->A)
2. No tests for concurrent cache access (despite using `RwLock<HashMap>`)
3. No tests for error recovery when filesystem errors occur mid-traversal
4. No tests for depth limit edge cases (exactly at limit, one over limit)

The code uses `ImportCache = Arc<RwLock<HashMap<PathBuf, Vec<Import>>>>` for shared state during parallel validation, but there are no concurrency tests.

## Suggested Tests
```rust
#[test]
fn test_circular_import_three_files() {
    // A imports B, B imports C, C imports A
}

#[test]
fn test_import_depth_exactly_at_limit() {
    // Chain of exactly MAX_IMPORT_DEPTH files
}

#[test]
fn test_concurrent_import_cache_access() {
    // Validate multiple files in parallel that share imports
}
```

## Effort
Medium

---
Found by `/audit-project` multi-agent review

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

- Add 11 new tests to the `ImportsValidator` inline test module covering gaps identified in issue #456
- No production code changes - purely additive test coverage

### Tests Added

**Multi-file circular chains**
- `test_cycle_detection_three_file_chain` - A→B→C→A (3-node cycle), verifies CC-MEM-002 fires and message contains the full cycle path
- `test_cycle_detection_four_file_chain` - A→B→C→D→A (4-node cycle), verifies CC-MEM-002 and cycle detection short-circuits before depth check
- `test_cycle_detection_three_file_chain_with_non_claude_root` - same 3-file cycle with SKILL.md root, verifies CC-MEM-002/003 are suppressed

**Depth boundary conditions**
- `test_depth_below_boundary_no_trigger` - 5-file chain (depth 4), well below MAX_IMPORT_DEPTH=5
- `test_depth_at_boundary_no_trigger` - 6-file chain (depth 5), exactly at MAX_IMPORT_DEPTH, verifies CC-MEM-003 does NOT fire

**Concurrent cache access**
- `test_concurrent_cycle_detection_no_deadlock` - 8 threads sharing `Arc<RwLock<HashMap>>` on a 2-file cycle
- `test_concurrent_three_file_cycle_shared_cache` - 10 threads on 3-file cycle, verifies cache populated for all 3 files
- `test_concurrent_cycle_near_depth_limit` - 5 threads on 5-file cycle at depth 4, verifies CC-MEM-002 fires and CC-MEM-003 does not
- `test_concurrent_diamond_dependency_no_duplicate_diagnostics` - diamond dependency (CLAUDE→{b,c}→shared→missing), asserts exactly 1 missing-import diagnostic per thread
- `test_concurrent_different_roots_shared_cache` - interleaved CLAUDE.md and SKILL.md threads sharing one cache; CLAUDE threads get CC-MEM-002, SKILL threads do not

**Mid-chain traversal**
- `test_missing_transitive_import_stops_traversal` - missing b.md in chain a.md→b.md stops recursion, c.md never referenced

## Test Plan

- [x] `cargo test --lib -p agnix-core -- rules::imports::tests --quiet` → 65/65 pass
- [x] `cargo test -p agnix-core` → full suite passes (2916 tests)
- [x] `cargo build -p agnix-core` → zero warnings

## Related Issues

Closes #456

### CODE REVIEW

File: crates/agnix-core/src/rules/imports.rs [hunk 1]
```diff
@@ -1801,4 +1801,427 @@ mod tests {
...
+        // 6 files, 5 hops deep - exactly at MAX_IMPORT_DEPTH = 5.
+        // Deepest point: depth=5, check 5+1=6 > 5 is true, so CC-MEM-003 WOULD fire...
+        // wait - at depth=4 we try to recurse into e (depth+1=5 <= 5, allowed).
+        // At depth=5 (visiting e) we try to recurse into leaf (depth+1=6 > 5, TRIGGERS).
+        // This test verifies the boundary: a chain reaching depth 5 (leaf at depth 5)
+        // is reachable but a further import from leaf would trigger CC-MEM-003.
+        // Since leaf has no further imports, CC-MEM-003 should NOT fire.
+        let temp = TempDir::new().unwrap();
+        let claude = temp.path().join("CLAUDE.md");
+        let a = temp.path().join("a.md");
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/imports.rs]: Fixed in commit 3b1ffab - rewrote the comment to match the actual chain (CLAUDE -> a -> b -> c -> d -> leaf) and describe the depth check precisely using the exact `depth + 1 > MAX_IMPORT_DEPTH` condition, removing the stale reference to file 'e'.

File: crates/agnix-core/src/rules/imports.rs [hunk 2]
```diff
@@ -1801,4 +1801,427 @@ mod tests {
...
+            })
+            .collect();
+
+        for handle in handles {
+            let diagnostics = handle
+                .join()
+                .expect("Thread panicked during cycle detection");
+            assert!(
+                diagnostics.iter().any(|d| d.rule == "CC-MEM-002"),
+                "Each thread should detect CC-MEM-002 in two-file cycle"
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/imports.rs]: Fixed in commit 3b1ffab - replaced the plain join() loop with a channel-based approach using mpsc::channel and recv_timeout(Duration::from_secs(10)). If a deadlock occurs, the test will now fail with a clear timeout message within 10 seconds rather than hanging the CI job indefinitely.

File: CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - **Confidence-tiered autofix engine**: `Fix` metadata now supports confidence, alternative groups, and dependencies; CLI adds `--fix-unsafe` and `--show-fixes`; core exposes confidence-based `FixApplyMode`/`FixApplyOptions`
 - **CI crate graph parity test**: New workspace-level test validates that all `Cargo.toml` workspace members are documented in CLAUDE.md, AGENTS.md, README.md, SPEC.md, and CONTRIBUTING.md - prevents architecture-doc drift
 - **`resolve_validation_root` file-input tests**: 7 integration tests covering single-file validation mode - validates file-input path behavior, unknown file type handling, project-level rule scoping, and nonexistent file edge case (#450)
+- **`ImportsValidator` concurrency and multi-file cycle tests**: 11 new tests covering thread-safety under concurrent validation, multi-file import cycles (3- and 4-file chains), depth boundary conditions (at, below, and above `MAX_IMPORT_DEPTH`), diamond dependency graphs, and mixed valid/invalid import scenarios (#456)
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply CHANGELOG.md]: Fixed in commit 3b1ffab - updated the CHANGELOG entry to say 'at and below MAX_IMPORT_DEPTH (complementing existing above-boundary coverage)' as suggested, since the above-boundary case is covered by the pre-existing test_depth_exceeded_in_claude_md and not newly added in this PR.
